src/alire/alire-solutions-diffs.adb

```diff
@@ -0,0 +1,221 @@
+         Trace.Info ("   New solution is valid.");
+      end if;
+
+      --  Early exit if no changes
+
+      if not This.Contains_Changes then
+         Trace.Info ("   No changes between former an new solution.");
+         return;
+      end if;
+
+      --  Detailed changes otherwise
+
+      for I in This.Changes.Iterate loop
+         declare
+            Former : Crate_Status renames This.Changes (I).Former;
+            Latter : Crate_Status renames This.Changes (I).Latter;
+         begin
+            if not Changed_Only or else Former /= Latter then
+
+               --  Show icon of change
+
+               Table.Append
+                 ("   "
+                  & (case This.Change (Key (I)) is
+                       when Added      => "✓",
+                       when Removed    => "✗",
+                       when External   => "↪",
+                       when Upgraded   => "⭧",
+                       when Downgraded => "⭨",
+                       when Unchanged  => "=",
+                       when Unsolved   => "⚠"));
+
+               --  Always show crate name
+
+               Table.Append (+Key (I));
+
+               --  Show most precise version available
+
+               if Latter.Status in Hinted | Needed then
+                  Table.Append (Best_Version (Latter));
+               else
+                  Table.Append (Best_Version (Former));
+               end if;
+
+               --  Finally show an explanation of the change depending on
+               --  status changes.
+
+               Table.Append
+                 ("("
+                  & (case This.Change (Key (I)) is
+                       when Added      => "new",
+                       when Removed    => "removed",
+                       when External   => "external",
+                       when Upgraded   => "upgraded from "
+                                          & Semver.Image (Former.Version),
+                       when Downgraded => "downgraded from "
+                                          & Semver.Image (Former.Version),
+                       when Unchanged  => "unchanged",
+                       when Unsolved   => "missing")
+                  & ")");
+
+               Table.New_Row;
+            end if;
+         end;
+      end loop;
+
+      Table.Print (Level => Info);
+   end Print;
+
+   -----------------------
+   -- Print_And_Confirm --
+   -----------------------
+
+   function Print_And_Confirm
+     (This         : Diff;
+      Changed_Only : Boolean)
+      return Boolean
+   is
+      use Utils.User_Input;
+   begin
+      if This.Contains_Changes then
+         Trace.Info ("Changes to dependency solution:");
+         Trace.Info ("");
+         This.Print (Changed_Only => Changed_Only);
+         Trace.Info ("");
+      else
+         Trace.Info
+           ("There are no changes between the former and new solution.");
+      end if;
+
+      return Utils.User_Input.Query
```

Comment:
I don't think the query belongs here. We might want to print the diff without asking this specific question.

Reply:
Right, there is a `Print` that does only that for these cases. This is to avoid repetition of the same prompt in `alr with`, `pin` and `update`. Still this one can be moved elsewhere to keep this class free from specifics.